protocols/gossipsub/src/protocol.rs

```diff
@@ -0,0 +1,343 @@
+                    message_ids: ihave.take_messageIDs().into_vec(),
+                }
+            })
+            .collect();
+
+        let iwant_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_iwant()
+            .into_iter()
+            .map(|mut iwant| {
+                GossipsubControlAction::IWant {
+                    // TODO: Potentially format the message ids better
+                    message_ids: iwant.take_messageIDs().into_vec(),
+                }
+            })
+            .collect();
+
+        let graft_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_graft()
+            .into_iter()
+            .map(|mut graft| GossipsubControlAction::Graft {
+                topic_hash: TopicHash::from_raw(graft.take_topicID()),
+            })
+            .collect();
+
+        let prune_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_prune()
+            .into_iter()
+            .map(|mut prune| GossipsubControlAction::Prune {
+                topic_hash: TopicHash::from_raw(prune.take_topicID()),
+            })
+            .collect();
+
+        control_msgs.extend(ihave_msgs);
+        control_msgs.extend(iwant_msgs);
+        control_msgs.extend(graft_msgs);
+        control_msgs.extend(prune_msgs);
+
+        Ok(Some(GossipsubRpc {
+            messages,
+            subscriptions: rpc
+                .take_subscriptions()
+                .into_iter()
+                .map(|mut sub| GossipsubSubscription {
+                    action: if sub.get_subscribe() {
+                        GossipsubSubscriptionAction::Subscribe
+                    } else {
+                        GossipsubSubscriptionAction::Unsubscribe
+                    },
+                    topic_hash: TopicHash::from_raw(sub.take_topicid()),
+                })
+                .collect(),
+            control_msgs,
+        }))
+    }
+}
+
+/// A message received by the gossipsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct GossipsubMessage {
+    /// Id of the peer that published this message.
+    pub source: PeerId,
+
+    /// Content of the message. Its meaning is out of scope of this library.
+    pub data: Vec<u8>,
+
+    /// A random sequence number.
+    pub sequence_number: Vec<u8>,
+
+    /// List of topics this message belongs to.
+    ///
+    /// Each message can belong to multiple topics at once.
+    pub topics: Vec<TopicHash>,
+}
+
+impl GossipsubMessage {
+    /// Converts message into a message_id.
+    // To be compatible with the go implementation
+    pub fn id(&self) -> String {
+        let mut source_string = self.source.to_base58();
+        // the sequence number should be a big endian uint64 (as per go implementation)
+        // avoid a potential panic by setting the seqno to 0 if it is not long enough.
+        let seqno = {
+            if (self.sequence_number.len() == 0) | (self.sequence_number.len() > 8) {
```

Comment:
```suggestion
            if (self.sequence_number.len() == 0) | (self.sequence_number.len() < 8) {
```

Reply:
Rereading this again, I think I misunderstood the intention. Let me take a look at how the go implementation handles this first.